kv/bulk: add metrics for bulk adder throughput

rafiss opened the pull request:
This change adds metrics to help diagnose the performance
of index backfills and other bulk operations that use the BufferingAdder.

Previously, operators had limited visibility into the progress and
performance of long-running bulk operations. During customer escalations
involving large index backfills (e.g., 230TB), it was difficult to
determine why operations were proceeding slowly despite available system
resources (CPU, I/O).

This change introduces two new metrics:
- bulk_adder.bytes_in: Tracks the rate of data entering the bulk adder
- bulk_adder.sst_count: Counts the number of SSTs generated

These metrics complement the existing IngestionPerformanceStats by
providing real-time visibility that can be monitored through standard
observability tools, helping operators answer questions like:
- What is the current throughput of the bulk operation?
- How many SSTs are being generated (and thus their average size)?
- Is the bulk operation making consistent progress?

Epic CRDB-42901
Release note (ops change): Added two new metrics for monitoring bulk
operations like index backfills: bulk_adder.bytes_in tracks the rate
of data being processed, and bulk_adder.sst_count tracks the number of
SST files generated. These metrics provide better observability into the
progress and performance of long-running bulk operations.

rafiss committed:
kv/bulk: add metrics for bulk adder throughput

This change adds observability metrics to help diagnose the performance
of index backfills and other bulk operations that use the BufferingAdder.

Previously, operators had limited visibility into the progress and
performance of long-running bulk operations. During customer escalations
involving large index backfills (e.g., 230TB), it was difficult to
determine why operations were proceeding slowly despite available system
resources (CPU, I/O).

This change introduces two new metrics:
- bulk_adder.bytes_in: Tracks the rate of data entering the bulk adder
- bulk_adder.sst_count: Counts the number of SSTs generated

These metrics complement the existing IngestionPerformanceStats by
providing real-time visibility that can be monitored through standard
observability tools, helping operators answer questions like:
- What is the current throughput of the bulk operation?
- How many SSTs are being generated (and thus their average size)?
- Is the bulk operation making consistent progress?

Fixes #140636

Release note (ops change): Added two new metrics for monitoring bulk
operations like index backfills: bulk_adder.bytes_in tracks the rate
of data being processed, and bulk_adder.sst_count tracks the number of
SST files generated. These metrics provide better observability into the
progress and performance of long-running bulk operations.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/140636)
<!-- Reviewable:end -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140636)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @jbowens, @kev-cao, and @sumeerbhola)

___
*[`pkg/kv/bulk/bulk_metrics.go` line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140636#-OIS9fODAn9wlMba1wVo:-OIS9fODAn9wlMba1wVp:b-tfrm0a) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2a84d460a114ba80dcb9c1f5de3be30d72d6634/pkg/kv/bulk/bulk_metrics.go#L44)):*
> ```Go
> 	metaBulkBytesIn = metric.Metadata{
> 		Name:        "bulk_adder.bytes_in",
> 		Help:        "Bytes being sent into the bulk adder",
> ```

I'm not sure what user-facing terminology we use here. (And also same for the release note.)


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
> Reviewed 4 of 4 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140636)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @annrpom, @jbowens, @kev-cao, and @rafiss)

___
*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/140636#-OIVu3tj66SsIdwZEpf-:-OIVu3tj66SsIdwZEpf0:bg7igkv) (no related file):*
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 
___
*[`pkg/kv/bulk/bulk_metrics.go` line 52 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/140636#-OIVolXf84x2V2JH6tn_:-OIVolXf84x2V2JH6tna:b1sqb7i) ([raw file](https://github.com/cockroachdb/cockroach/blob/db739c5243c0d593e40d671d01da737d3948e5b6/pkg/kv/bulk/bulk_metrics.go#L52)):*
> ```Go
> 		Help:        "Bytes being flushed out of the bulk adder",
> 		Measurement: "Bytes flushed",
> 		Unit:        metric.Unit_BYTES,
> ```

I'm trying to step back and understand how we will use these metrics when trying to understand the behavior of an ongoing backfill (or import or whatever uses BufferingAdder).
- We can of course see the aggregate rate across the cluster of k-v pairs being buffered and sent. We could also look at the per node rate and see if the work of generating these ssts is equally distributed across the cluster.
- I'm not sure having two different metrics for bytes_in and bytes_out helps. The buffering is presumably short-lived (for a few seconds), so I suspect the rate will effectively look the same.
- It may be worth counting the number of ssts generated. It would give us the mean size that is being generated. The mean will be in terms of uncompressed key-value bytes.
- One thing we were missing in that investigation is a counter of the total byte size of addsstables received by a store. See the speculation in https://cockroachlabs.slack.com/archives/C08AS0GMK5J/p1738697660886349?thread_ts=1738606046.586139&cid=C08AS0GMK5J. That will require instrumenting `kvserver.Store`. Probably in `Store.maybeThrottleBatch` since it already switches on the request type and has a case for `*kvpb.AddSSTableRequest`. That can be a separate PR.

What am I missing? Sometimes it helps to run this, i.e., run an index backfill with these changes and see what questions it answers, and what it doesn't, before merging.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140636)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @annrpom, @jbowens, @kev-cao, and @sumeerbhola)

___
*[`pkg/kv/bulk/bulk_metrics.go` line 52 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/140636#-OIVolXf84x2V2JH6tn_:-OIYLuuE8nHe7UOgcgYQ:b9v9pkc) ([raw file](https://github.com/cockroachdb/cockroach/blob/db739c5243c0d593e40d671d01da737d3948e5b6/pkg/kv/bulk/bulk_metrics.go#L52)):*
> I'm not sure having two different metrics for bytes\_in and bytes\_out helps. The buffering is presumably short-lived (for a few seconds), so I suspect the rate will effectively look the same.

I wanted to make sure I understood -- the counter I added for `bytesout` is incremented when the bytes are flushed out of the `SSTBatcher` underneath the `BufferingAdder`. I wasn't sure if the buffering you refer to is about the time spent in the `SSTBatcher`, or in the internal buffers in `BufferingAdder`, or both combined.

> It may be worth counting the number of ssts generated. It would give us the mean size that is being generated. The mean will be in terms of uncompressed key-value bytes.

That sounds like a good idea, I will definitely add that in this PR.

> One thing we were missing in that investigation is a counter of the total byte size of addsstables received by a store.

I'm not familiar with the code where I'd add that.  Is it something that the Storage team could work on? If not, could you show me where to add it?

> Sometimes it helps to run this, i.e., run an index backfill with these changes and see what questions it answers, and what it doesn't, before merging.

I'll do this once I've added the SSTs generated metric.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140636)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @annrpom, @jbowens, @kev-cao, and @rafiss)

___
*[`pkg/kv/bulk/bulk_metrics.go` line 52 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/140636#-OIVolXf84x2V2JH6tn_:-OI_zBVl7j6VaDM2jP55:b-8xg8tt) ([raw file](https://github.com/cockroachdb/cockroach/blob/db739c5243c0d593e40d671d01da737d3948e5b6/pkg/kv/bulk/bulk_metrics.go#L52)):*
> I wanted to make sure I understood -- the counter I added for `bytesout` is incremented when the bytes are flushed out of the `SSTBatcher` underneath the `BufferingAdder`. I wasn't sure if the buffering you refer to is about the time spent in the `SSTBatcher`, or in the internal buffers in `BufferingAdder`, or both combined.

I meant both. My understanding (based on a cursory examination of the code) was that most of the buffering happens in `BufferingAdder`. When it is full, we have to flush it. At that point the key-value pairs are fed in sorted order to the `SSTBatcher` (since SSTs hold sorted keys) - all of this happens in `BufferingAdder.doFlush`. The `SSTBatcher` is buffering the sst it is generating, and will internally flush (within the context of being exercised by `BufferingAdder.doFlush`) when the sst is large enough or range boundaries are crossed (and will inject a possibly small `bulkio.ingest.flush_delay`).


<!-- Sent from Reviewable.io -->

dt commented:
Similar question to @sumeerbhola: how are these metrics useful to or actionable by the operator of a cluster? What alerts do we see an operator defining on these metrics and/or how would they use these metrics to configure or operate their cluster?

rafiss commented:
I still plan to return to this PR (and was planning to add more details to the commit message / release note at that point).

> how are these metrics useful to or actionable by the operator of a cluster? What alerts do we see an operator defining on these metrics and/or how would they use these metrics to configure or operate their cluster?

We don't expect anyone to define alerts based on these metrics. The reason they can be useful is to help answer questions about index backfills. This PR is being worked on in response to a customer escalation where they were doing a 230 TB index backfill (https://github.com/cockroachlabs/support/issues/3169 and https://cockroachlabs.slack.com/archives/C08AS0GMK5J/p1738692845947149?thread_ts=1738606046.586139&cid=C08AS0GMK5J), and lacked observability into why the backfill was taking as long as it was. The main question the operators of that cluster wanted to answer was: "it seems like the backfill is going slower than we want, why is that?" While trying to help them, the main question we found ourselves asking "why does the cluster seemingly have extra resource capacity (CPU and I/O weren't too high), yet the backfill is still going slow?" By answering these questions, an operator of a cluster would know more about what they may need to change in order to make the backfill go faster.

dt commented:
I ask since I'm curious how this complements or overlaps with the existing SST-batcher/buffering-adder instrumentation we have via the structured `IngestionPerformanceStats` reporting, particularly as we've been talking about how we need to rein in the proliferation of non-alerting/non-actionable metrics.

rafiss commented:
This is RFAL. I have added the SST Count metric as suggested, and added details in the commit message to provide context about how these metrics can be useful. Here's an example of them from an index backfill test I ran:
![CleanShot 2025-05-29 at 16 27 07@2x](https://github.com/user-attachments/assets/6a4032f0-ca57-4454-a36d-58135f489a37)

sumeerbhola reviewed (approved):
> Reviewed 2 of 5 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140636)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @annrpom, @jbowens, @kev-cao, @rafiss, and @yuzefovich)

___
*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/140636#-OIVu3tj66SsIdwZEpf-:-OSdWFqY86_hxbnDiXAe:bg7igkv) (no related file):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>
</blockquote></details>

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 
___
*[`pkg/kv/bulk/bulk_metrics.go` line 50 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/140636#-OSdVDjd1vsPCp6HpSR5:-OSdVDjd1vsPCp6HpSR6:b-cz021) ([raw file](https://github.com/cockroachdb/cockroach/blob/c34364487a9bf9a90473edd507ca6e7964738520/pkg/kv/bulk/bulk_metrics.go#L50)):*
> ```Go
> 	metaBulkSSTCount = metric.Metadata{
> 		Name:        "bulk_adder.sst_count",
> 		Help:        "Number of SSTs generated by the bulk adder",
> ```

This count is incremented after the AddSSTable request has executed. So it includes delays incurred in BatchRequest execution at the remote nodes (which includes AC and latching delays). Does that provide enough information for debugging? Or would you want to export multiple cumulative durations, like the time spent doing work in BufferingAdder+SSTBatcher as one metric, and the time spent waiting for BatchRequest execution as another metric. That would give us an idea of where the slowness is.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 5 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140636#-:-OSewAZ-8oRU1GL9tXFA:bnfp4nl)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @annrpom, @jbowens, @jeffswenson, @kev-cao, @rafiss, and @sumeerbhola)

___
*[`pkg/kv/bulk/bulk_metrics.go` line 52 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/140636#-OIVolXf84x2V2JH6tn_:-OSepnwJ33kCN4r5ZYT8:b-shcrx6) ([raw file](https://github.com/cockroachdb/cockroach/blob/db739c5243c0d593e40d671d01da737d3948e5b6/pkg/kv/bulk/bulk_metrics.go#L52)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

> I wanted to make sure I understood -- the counter I added for `bytesout` is incremented when the bytes are flushed out of the `SSTBatcher` underneath the `BufferingAdder`. I wasn't sure if the buffering you refer to is about the time spent in the `SSTBatcher`, or in the internal buffers in `BufferingAdder`, or both combined.

I meant both. My understanding (based on a cursory examination of the code) was that most of the buffering happens in `BufferingAdder`. When it is full, we have to flush it. At that point the key-value pairs are fed in sorted order to the `SSTBatcher` (since SSTs hold sorted keys) - all of this happens in `BufferingAdder.doFlush`. The `SSTBatcher` is buffering the sst it is generating, and will internally flush (within the context of being exercised by `BufferingAdder.doFlush`) when the sst is large enough or range boundaries are crossed (and will inject a possibly small `bulkio.ingest.flush_delay`).
</blockquote></details>

Your understanding matches mine.

There are three types of activities happening:
- first, we need to construct KVs to ingest. This is done in the user of BulkAdder, probably always in another goroutine (often multiple goroutines)
- next, the user of BulkAdder ingests the KVs which are buffered in the BulkAdder.
- in the same goroutine, when enough data has been buffered, the BulkAdder will start generating SSTs via `SSTBatcher.AddMVCCKey` and eventually the BulkAdder will flush SSTBatcher. (SSTBatcher can also decide to flush on its own, which the BulkAdder will only be aware of via `onFlush` callback.)

It sounds like Sumeer suggests that duration metrics for some of these "stages" would be helpful. We already have `IngestionPerformanceStats.FlushWait` which tracks the third stage. There are also some other duration counters in `IngestionPerformanceStats` at different granularity. IIUC these perf stats don't show up as metrics, only in the logs and in the trace, and maybe on the jobs page (?).

It seems hard to provide a cluster-wide metric for these durations. The additions in this PR seem useful. I'm curious whether @jeffswenson has some thoughts on what cluster-level (i.e. that can be viewed via DataDog) observability we can add here.
___
*[`-- commits` line 24 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/140636#-OSenBmM3cylFYiCAy3c:-OSenBmM3cylFYiCAy3d:b-3tb5d8):*
nit: this is linking to the PR where this commit is.


<!-- Sent from Reviewable.io -->

jeffswenson commented:
I think we should cut the `bulk_adder.sst_count` metric. It is basically a clone of cockroachdb.distsender.rpc.addsstable.sent, but measured at a different level. 

My preferred way of measuring throughput would be to add a per rpc throughput metric to distsender. With the new labeled metrics, that would allow us to measure add sst throughput from the client and would be a generally useful debugging tool. But I acknowledge this is more work, so its not a requirment.

I'm currently working on adding some async flushing and pipelining back to the sst batcher. Once my change is live, I think the metric we would want from the sst batcher is time spent waiting on flushes.

rafiss commented:
> I think we should cut the bulk_adder.sst_count metric. It is basically a clone of cockroachdb.distsender.rpc.addsstable.sent, but measured at a different level.

Here is a chart of both metrics from a test I just ran:
![CleanShot 2025-06-24 at 16 58 58@2x](https://github.com/user-attachments/assets/a4a70456-c2e1-4676-8e85-e18771193d16)

They don't have the exact same shape, but are generally similar. They also have different magnitudes. This test has a foreground workload running, but I wouldn't expect that workload to be causing any AddSST commands, so I don't know how to explain the differences between the charts. Do you have an explanation?

>  I think the metric we would want from the sst batcher is time spent waiting on flushes.

That makes sense to add. I can work on that.

jeffswenson commented:
>They also have different magnitudes. This test has a foreground workload running, but I wouldn't expect that workload to be causing any AddSST commands, so I don't know how to explain the differences between the charts. Do you have an explanation?

There are three layers to the buffering adder used by index builds:
- buffering adder: collects unsorted keys, sorts them, then flushes them. This is expected to grow to be 512 Mib flushes (schemachanger.backfiller.max_buffer_size).
- sst batcher: collects sorted keys and flushes them based on size or range boundaries. If this index is out of order, then ranges are going to be responsible for most fragmentation. If its in order, the limiting factor is the max sst request size which is 16 MiB (kv.bulk_ingest.batch_size).
- sst adder: flushes a single logical sst. This may be retried in the case of errors or fragmented if there is an unexpected range boundary. There will be an unexpected range boundary if a range splits or if the sst batcher was not configured with a range cache.

This change is measuring flushes at the buffering adder level. The RPC metric is measuring flushes at the RPC layer, which will match the number of SSTs flushed by the sst adder.

rafiss commented:
This has gone stale, and things have changed with the new distributed merge implementation too
